### ISSUE TITLE
How to calculate the loss for imperfect CSI case?  

### ISSUE THREAD

**Issue by SharonWxy** [TianLin0509/beamforming-design-with-DL-approach#2]
First I’d like to thank you very much for providing the code online for further learning. Then, if you get some time, may I ask some questions regarding your code? I put my question here as following:

1. For imperfect CSI case, to calculate the loss, do you use the true H, or the estimated H? I think it should be the true H, right? 
2. If you use the real H, how do you pass the value of true H to the self-defined loss function? 

I really appreciate your reply, and thank you very much in advance.

**Comment by TianLin0509** [comment]
1. For imperfect CSI case, we calculate the loss with the true H during the offline training. (Noticed when deployed online, it is not needed).
2. In this code, we use the input layer of Keras to input the true H, which is directly used to compute the loss. (Only used to compute loss and did not through the neural network, so when deployed online, i.e., the testing procedure, only imperfect CSI is needed to fed into the BFNN to compute the analog beamformer.
3. Thank you for your approval of this work.

**Comment by SharonWxy** [comment]
Thank you very much.~

**Comment by AlirezaJahanbakhshi** [comment]
Dear Mr. Lin,
I also have question about calculation of loss function. would you please tell where did you exactly calculate loss function in your code? 
I appreciate your kind cooperation

**Comment by TianLin0509** [comment]
The calculation of loss function, is discussed in the issues of the repo. You can refer to it, because some other people have the similar question, so I answered it already. The channels is generated via the codes provided via the author of "Reliable Beamspace Channel Estimation for Millimeter-Wave Massive MIMO Systems with Lens Antenna Array", you can find it in the authors'own website. I just used his codes. 


You can also try other channel models, hopefully my BFNN can also work. 


-----原始邮件-----
发件人:AlirezaJahanbakhshi <notifications@github.com>
发送时间:2020-02-25 17:48:57 (星期二)
收件人: TianLin0509/BF-design-with-DL <BF-design-with-DL@noreply.github.com>
抄送: "Lin Tian" <17210720082@fudan.edu.cn>, Comment <comment@noreply.github.com>
主题: Re: [TianLin0509/BF-design-with-DL] How to calculate the loss for imperfect CSI case? (#2)



Dear Mr. Lin,
I also have question about calculation of loss function. would you please tell where did you exactly calculate loss function in your code?
I appreciate your kind cooperation

**Comment by AlirezaJahanbakhshi** [comment]
Thank you very much for your answer.
